paddle/fluid/platform/monitor.h

```diff
@@ -0,0 +1,182 @@
+
+  static StatRegistry<T>& Instance() {
+    static StatRegistry<T> r;
+    return r;
+  }
+  StatValue<T>* get(const std::string& name) {
+    std::lock_guard<std::mutex> lg(mutex_);
+    auto it = stats_.find(name);
+    if (it != stats_.end()) {
+      return it->second;
+    } else {
+      return nullptr;
+    }
+  }
+  int add(const std::string& name, StatValue<T>* stat) {
+    std::lock_guard<std::mutex> lg(mutex_);
+    auto it = stats_.find(name);
+    if (it != stats_.end()) {
+      return -1;
+    }
+    stats_.insert(std::make_pair(name, stat));
+    return 0;
+  }
+
+  void publish(std::vector<ExportedStatValue<T>>& exported,  // NOLINT
+               bool reset = false) {
+    std::lock_guard<std::mutex> lg(mutex_);
+    exported.resize(stats_.size());
+    int i = 0;
+    for (const auto& kv : stats_) {
+      auto& out = exported.at(i++);
+      out.key = kv.first;
+      out.value = reset ? kv.second->reset() : kv.second->get();
+    }
+  }
+
+  std::vector<ExportedStatValue<T>> publish(bool reset = false) {
+    std::vector<ExportedStatValue<T>> stats;
+    publish(stats, reset);
+    return stats;
+  }
+
+ private:
+  std::mutex mutex_;
+  std::unordered_map<std::string, StatValue<T>*> stats_;
+};
+
+}  // namespace platform
+}  // namespace paddle
+
+// Because we only support these two types in pybind
+#define REGISTER_FLOAT_STATUS(item) \
+  extern paddle::platform::StatValue<float> _##item;
+
+#define REGISTER_INT_STATUS(item) \
+  extern paddle::platform::StatValue<int64_t> _##item;
+
+#define STAT_ADD(item, t) _##item.increase(t)
+#define STAT_SUB(item, t) _##item.decrease(t)
+
+// Support add stat value by string
+#define STAT_INT_ADD(item, t) \
+  paddle::platform::StatRegistry<int64_t>::Instance().get(item)->increase(t)
+#define STAT_INT_SUB(item, t) \
+  paddle::platform::StatRegistry<int64_t>::Instance().get(item)->decrease(t)
+
+#define STAT_FLOAT_ADD(item, t) \
+  paddle::platform::StatRegistry<float>::Instance().get(item)->increase(t)
+#define STAT_FLOAT_SUB(item, t) \
+  paddle::platform::StatRegistry<float>::Instance().get(item)->decrease(t)
+
+#define STAT_RESET(item, t) _##item.reset(t)
+#define STAT_GET(item) _##item.get()
+
+#define DEFINE_FLOAT_STATUS(item)                    \
+  paddle::platform::StatValue<float> _##item(#item); \
+  int TouchStatRegistrar_##item() {                  \
+    _##item.Touch();                                 \
+    return 0;                                        \
+  }
+
+#define DEFINE_INT_STATUS(item)                        \
+  paddle::platform::StatValue<int64_t> _##item(#item); \
+  int TouchStatRegistrar_##item() {                    \
+    _##item.Touch();                                   \
+    return 0;                                          \
+  }
+
+#define USE_STAT(item)                    \
+  extern int TouchStatRegistrar_##item(); \
+  UNUSED static int use_stat_##item = TouchStatRegistrar_##item()
+
+REGISTER_INT_STATUS(STAT_total_feasign_num_in_mem)
```

Comment:
why not move these registration to where they need?

Reply:
Maybe for monitor stat, putting them together is more clear?